### ISSUE TITLE
[GDnative] Rethink `DEBUG_METHODS_ENABLED` to keep introspection in release

### ISSUE THREAD

**Issue by touilleMan** [godotengine/godot#15065]
I've realized a bit late that Godot introspection system consider method's arguments information (i.e. [arguments types and names](https://github.com/godotengine/godot/blob/37aab45091961e7c063e9c77d6c4835b0c14806a/core/method_bind.h#L196)) as only useful for debug (and so are totally [ignored when switching to release](https://github.com/godotengine/godot/blob/37aab45091961e7c063e9c77d6c4835b0c14806a/core/class_db.h#L77))

While this was true before, this is no longer the case now that the GDnative API provides introspection system to 3rd party module.
For instance, my Python binding lazily generate the binding to a Godot class by using the introspection API.
- argument name information improve readability of the generated method
- name information also allow to do key/value style argument passing (e.g. `camera.set_orthogonal(z_near=42.0)`)
- type information allow to do argument type checking before calling Godot. This is good because:
    1) it saves conversion time if type is wrong
    2) it allow to uses ptrcall to avoid converting arguments to Variant, having Godot doing an additional lookup, having to check return type for type error and convert it back to a readable thing for the caller

On top of that from a GDnative user, it is really strange to have this introspection API returns different things (even worse: the returned format is the same but informations differs !) depending on if you use the Godot binary with or without the editor.

**Comment by touilleMan** [comment]
@karroffel @reduz sorry to directly ping you this topic, but I think this is a serious one considering it implications and how deep we are into the release process of 3.0

**Comment by reduz** [comment]
It makes the binary more fatty, but you are free to make your own template
with tools or debug enabled

**Comment by touilleMan** [comment]
Thanks for the answer reduz ^^

> It makes the binary more fatty

Yes, from my tests I go from 25mo to 28mo when enabling `DEBUG_METHODS_ENABLED` by hand (on linux 64 with gcc-7).
This 10% change seems not much on PC, but I guess your concern is more about less powerful targets (especially the web one).

> but you are free to make your own template with tools or debug enabled

From my tests, introspection is provided on the editor and debug release version of Godot. That means I would "only" have to provide a release version with `DEBUG_METHODS_ENABLED` enabled.
While it seems doable, I would rather not doing this:
- supporting a custom build is more work on my side :'-(
- it would make a poor user experience to ask to manually download a custom release template
- it seems like dividing the community, the whole point of GDNative was to avoid having as many custom builds as needs...

To avoid this I see two solutions:
- add new target to the official release template (i.e. release with introspection), I guess @akien-mga wouldn't be found of this solution :trollface: 
- tell users it use the official release-debug template instead of the release when using introspection. This is the *status quo* solution, however I don't know the differences between this version and the release (except for the introspection). From what I saw it only enable the `DEBUG_ENABLED` which provide better warning prints this. This seems like a really small change (I would even say this should be the default release mode) so I feel like I'm missing something there...


Regarding my initial post, my biggest concern was about the correctness of the API toward 3rd party users.
For instance using `_ClassDB.get_class_method` on `_OS` give us the following json for method `_OS.native_video_play` when introspection is available:
```
{
   'name':'native_video_play',
   'args':[
      {
         'name':'path',
         'class_name':'',
         'type':4,
         'hint':0,
         'hint_string':'',
         'usage':7
      },
      {
         'name':'volume',
         'class_name':'',
         'type':3,
         'hint':0,
         'hint_string':'',
         'usage':7
      },
      {
         'name':'audio_track',
         'class_name':'',
         'type':4,
         'hint':0,
         'hint_string':'',
         'usage':7
      },
      {
         'name':'subtitle_track',
         'class_name':'',
         'type':4,
         'hint':0,
         'hint_string':'',
         'usage':7
      }
   ],
   'default_args':[

   ],
   'flags':1,
   'id':583,
   'return':{
      'name':'',
      'class_name':'Error',
      'type':2,
      'hint':0,
      'hint_string':'',
      'usage':262151
   }
},
```
which becomes:
```
{
   'name':'native_video_play',
   'args':[

   ],
   'default_args':[

   ],
   'flags':1,
   'id':0,
   'return':{
      'name':'',
      'class_name':'',
      'type':0,
      'hint':0,
      'hint_string':'',
      'usage':7
   }
}
```

What I find fundamentally wrong here is there is Godot act just like everything went smooth while providing dummy data to the user.
So in the end the API is not consistent between dev and release modes, and failed silently on this inconsistency... And I feel like the first victim of this trick :smile: 

I guess a really simple way to go around this would be to add warning prints when using those function when introspection is disabled.

**Comment by akien-mga** [comment]
Fixed by #15337.
